File: TUM_Lens/app/src/main/java/com/maxjokel/lens/helpers/FreezeAnalyzer.java
package com.maxjokel.lens.helpers;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.media.Image;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageProxy;

/* + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + +

    13.09.2020, 19:45

    We use this custom CameraX Analyzer to display the last camera view frame as 'frozen'
    when the user pauses the classification.

    While there (were) are several approaches to this, this one is apparently the one suggested
    by the Google developer team.

    IDEA
     - adapted from here: https://stackoverflow.com/a/59674075
     - we use a callback interface to communicate between 'ViewFinder.java' and this class
     - here, an 'event' is fired   [Part 1]
     - it is then processed in 'ViewFinder.java', where we actually freeze the UI   [Part 2]

+ + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + + */

public class FreezeAnalyzer implements ImageAnalysis.Analyzer {


    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // init new Logger instance
    private static final Logger LOGGER = new Logger();


    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // init global variables
    private FreezeCallback callback;
    private boolean isFrozen;
    private int lens_front_back = 0; // [0 = back, 1 = front]; set 'back' as default;


    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // constructor
    public FreezeAnalyzer(FreezeCallback freezeCallback){
        this.isFrozen = false;
        this.callback = freezeCallback;
    }


    @Override
    public void analyze(@NonNull ImageProxy image) {

        // analyze() does nothing but closing the image;
        // if the 'isFrozen' flag is set, it will return a bitmap within its callback

        if (isFrozen){

            // convert the CameraX YUV ImageProxy to a (RGB?) bitmap
            @androidx.camera.core.ExperimentalGetImage
            Image img = image.getImage();

            @SuppressLint("UnsafeExperimentalUsageError")
            Bitmap bmp = ImageUtils.toBitmap(img, lens_front_back, image.getImageInfo().getRotationDegrees());

            // trigger new callback
            callback.onFrozenBitmap(bmp);

            // reset flag
            isFrozen = false;

        }

        // close image to continue processing the next frames
        image.close();
    }



    public void freeze(int lens){
        isFrozen = true;
        lens_front_back = lens;
    }

}

File: TUM_Lens/app/src/main/java/com/maxjokel/lens/helpers/FreezeAnalyzer.kt
package com.maxjokel.lens.helpers

import android.annotation.SuppressLint
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import com.maxjokel.lens.helpers.ImageUtils.toBitmap

/* We use this custom CameraX Analyzer to display the last camera view frame as 'frozen'
   when the user pauses the classification. While there (were) are several approaches to this,
   this one is apparently the one suggested by the Google developer team.

   IDEA
    - adapted from here: https://stackoverflow.com/a/59674075
    - we use a callback interface to communicate between 'ClassificationActivity' and this class
    - here, an 'event' is fired [Part 1]
    - it is then processed in 'ClassificationActivity' where we actually freeze the UI [Part 2] */

class FreezeAnalyzer(private val callback: FreezeCallback) : ImageAnalysis.Analyzer {

    private var isFrozen = false
    private var lensFrontBack = 0 // [0 = back, 1 = front]; set 'back' as default;

    override fun analyze(imgProxy: ImageProxy) {
        // analyze() does nothing but close the image
        // if the 'isFrozen' flag is set, it will return a bitmap within its callback
        if (isFrozen) {
            // convert the CameraX YUV ImageProxy to a (RGB?) bitmap
            @ExperimentalGetImage val img = imgProxy.image
            @SuppressLint("UnsafeExperimentalUsageError")
            val bmp = toBitmap(img!!, lensFrontBack, imgProxy.imageInfo.rotationDegrees)
            callback.onFrozenBitmap(bmp) // trigger new callback
            isFrozen = false // reset flag
        }
        imgProxy.close() // close image proxy to continue processing the next frames
    }

    fun freeze(lens: Int) {
        isFrozen = true
        lensFrontBack = lens
    }

    companion object {
        private val LOGGER = Logger()
    }
}